src/accelerator.rkt

```diff
@@ -0,0 +1,127 @@
+#lang racket
+
+(require
+ math/bigfloat
+
+ "compiler.rkt"
+ "ground-truth.rkt"
+ (submod "syntax/syntax.rkt" internals)
+ (submod "syntax/rules.rkt" internals)
+ "syntax/types.rkt")
+
+(provide
+ register-accelerator-operator!
+ register-accelerator-implementation!
+ expand-accelerators)
+
+(module+ internals
+  (provide register-accelerator-operator!
+           register-accelerator-implementation!))
+
+;; This might have to stay, it's from common.rkt
+(define (sym-append . args)
+  (string->symbol (apply string-append (map ~a args))))
+
+(struct accelerator-operator (body variables itypes otypes))
+
+(define accelerator-operators (make-hasheq))
+
+(define (register-accelerator-operator!
+         name body variables
+         [itypes (make-list (length variables) 'real)] [otype 'real])
+  (define ruleset-name (sym-append name '- 'accelerator))
+  (define define-name (sym-append name '- 'define))
+  (define undefine-name (sym-append name '- 'undefine))
+  (hash-set! accelerator-operators
+             name
+             (accelerator-operator body variables itypes otype))
+  (register-operator! name
+                      itypes otype
+                      (list (cons 'ival (compile-spec body variables))))
+  ;; TODO: Are the groups right now?
+  (register-ruleset*! ruleset-name
+                     '(numerics simplify)
+                     (map cons variables itypes)
+                     `((,define-name ,body (,name ,@variables))
+                       (,undefine-name (,name ,@variables) ,body))))
+
+(define (register-accelerator-implementation! operator name
+                                              itypes otype
+                                              [implementation #f])
+  (match-define (accelerator-operator body variables _ _) (dict-ref accelerator-operators operator))
+  (register-operator-impl!
+   operator name
+   itypes otype
+   (list
+    (cons 'fl
+          (or 
+           implementation
+           (compose first (eval-progs-real 
+                           (list body)
+                           (list (context variables otype itypes)))))))))
+
```

Comment:
I'm assuming you'll add the macros? Check `src/syntax/syntax.rkt` for inspiration!